Shared/Realm.Shared/ISchemaObject.cs

```diff
@@ -0,0 +1,37 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using Realms.Schema;
+
+namespace Realms
+{
+    /// <summary>
+    /// An object describing its properties in terms of a <see cref="ObjectSchema"/>.
+    /// </summary>
+    public interface ISchemaObject
```

Comment:
Yes, TBH, the hardest part of this PR was coming up with a proper name for that :/ I hate `ISchemaObject`, but the alternatives I could think of were `ISchemableObject`, `IObjectWithSchema` and variations of those which were even worse.